### DISCUSSION TITLE
PSD/PSB - time of reading details about file compression

### DISCUSSION THREAD

**Discussion by flpxra** [ImageMagick/ImageMagick#4809] [Help]
I use magick `identify` tool to read whether the PSD / PSB file is compressed or not.
Overall, everything works, but reading a file that is about 1.5 GB and contains several layers takes over 20 seconds.
My guess is that this is because ImageMagick processes all layers (as seen below `composite/load` events) as images
Is there any option to skip all file processing and read only the data that interests me?

[![image.png](https://i.postimg.cc/K8TsTy7M/image.png)](https://postimg.cc/jDsXVmFx)

**Comment by fmw42** [comment]
What data interests you?  You can read just the first layer if you want via

magick identify -verbose image.psb[0]

Or you can get some parameter using

magick image.psb[0] -format "%[fx:some_parameter]" info:

See https://imagemagick.org/script/escape.php

**Reply by flpxra** [reply]
I am interested in compression and I know how to get it: `identify.exe -format "%f-%[compression]\n"  IMG_2708.psb`
Photoshop works as follows: when you save a PSD / PSB file without compression, several layers get compressed anyway but most none. When I save a file with compression, the first layer is uncompressed but all the others are.
So the surest way to know if a file was saved with or without compression is to read the compression from all layers, which takes an awfully long time.

Additional question: how to read something from the last layers without knowing the number of layers in the file?

**Reply by fmw42** [reply]
Last layer is [-1]. Second last is [-2] etc

**Reply by flpxra** [reply]
I does not work it returns always: `undefined `no matter if I use [-1] or [-9999]

**Reply by fmw42** [reply]
try  convert rather than identify

`convert image.suffix[-1] -format "%f-%[compression]\n" info:
`

**Reply by flpxra** [reply]
Ok, choosing one layer to process works on some files and then I get the information quickly but on others, despite selecting one layer `IMG_3045.psb[1]`, ImageMagick recalculates all of them but returns result for only one, selected. Do you know why this can be so?

```
r:\>identify.exe -monitor -format "%f-%[compression]\n" IMG_3045.psb[1]
Composite/Image[IMG_3045.psb]: 5199 of 5200, 100% complete
Load/Images[IMG_3045.psb]: 37 of 38, 100% completeomplete
Composite/Image[IMG_3045.psb]: 5199 of 5200, 100% complete
Composite/Image[IMG_3045.psb]: 5199 of 5200, 100% complete
Composite/Image[IMG_3045.psb]: 5199 of 5200, 100% complete
Composite/Image[IMG_3045.psb]: 5199 of 5200, 100% complete
Composite/Image[IMG_3045.psb]: 5199 of 5200, 100% complete
Composite/Image[IMG_3045.psb]: 5199 of 5200, 100% complete
Composite/Image[IMG_3045.psb]: 5199 of 5200, 100% complete
Composite/Image[IMG_3045.psb]: 5199 of 5200, 100% complete
Composite/Image[IMG_3045.psb]: 5199 of 5200, 100% complete
Composite/Image[IMG_3045.psb]: 5199 of 5200, 100% complete
Composite/Image[IMG_3045.psb]: 5199 of 5200, 100% complete
Composite/Image[IMG_3045.psb]: 5199 of 5200, 100% complete
Composite/Image[IMG_3045.psb]: 5199 of 5200, 100% complete
Composite/Image[IMG_3045.psb]: 5199 of 5200, 100% complete
Composite/Image[IMG_3045.psb]: 5199 of 5200, 100% complete
Composite/Image[IMG_3045.psb]: 5199 of 5200, 100% complete
Composite/Image[IMG_3045.psb]: 5199 of 5200, 100% complete
Merge/Layers[IMG_3045.psb]: 16 of 17, 100% complete
IMG_3045.psb-Zip
```

**Reply by fmw42** [reply]
Sorry, no.  Can you link to one of your PSB files that has issues like that if it is not huge.

**Reply by fmw42** [reply]
Is it a group layer or have a lot of linked layers?

**Reply by flpxra** [reply]
I put a file here that I have cropped to take up less space: [https://drive.google.com/file/d/1_snfV_7YI3EI7inSV8zqPqZaodb9qO6g/view?usp=sharing](https://drive.google.com/file/d/1_snfV_7YI3EI7inSV8zqPqZaodb9qO6g/view?usp=sharing)
I use this command: `identify.exe -monitor -format "%f-%[compression]\n" new.psb[1]`

It doesn't take long on this file, but on my original file it's over 20 seconds.

**Reply by fmw42** [reply]
Adobe wants an account to view that file.  Please post one that is not private.

**Reply by flpxra** [reply]
How private, where do you see such a message? It's a normal PSB file, you download it to disk and open it in Photoshop.

**Reply by fmw42** [reply]
Photoshop keeps asking me for an account

**Reply by flpxra** [reply]
I don't know what that means but I didn't know that you can use PS without an adobe account. My file doesn't contain anything fancy. I think this is a problem with a particular Photoshop installation, not with the file.

**Reply by fmw42** [reply]
I have been using PS for several years via cloud and have never had to provide a log-in to view a file on my system. So I am puzzled. I will try again.

**Reply by fmw42** [reply]
I logged into my account again and got past that. Your file has many linked layers. But when I do

`identify -format "%f-%[compression]\n" new.psb.psd[1]`

new.psb.psd-Zip

I just get the above one line.  So I cannot reproduce your issue. I am using IM 6.9.12-38 and the same result on IM 7.1.0-23. 

What is your IM version and platform/OS?

**Reply by flpxra** [reply]
Use -monitor parameter to see that ImageMagick goes thru composition/loading every layer. When image resolution is big it takes time about 20 seconds, no matter I request one layer or all.

**Reply by fmw42** [reply]
I do not know why it would be slower other than -monitor has to print something every so often. But this is a question for clarification from the IM developers.

**Reply by fmw42** [reply]
Why do you have to use -monitor?  That is not necessary when using identify.
